### PR TITLE
FIX: Processors are now always applied when reading the value of an action.

### PR DESCRIPTION
**Pull request by andrew-oc**
Case [1293728](https://issuetracker.unity3d.com/product/unity/issues/guid/1293728/)([Fogbugz](https://fogbugz.unity3d.com/f/cases/1293728/))

### Description

The original bug report suggested that the dead zone processor was interacting with the normalize processor and causing it to malfunction. The actual issue was that calling InputAction.ReadValue() when no binding/control is actuated always returns default(T), where in this case the user had a normalize processor on a stick X axis binding that they wanted to return 0.5 at the default state, instead of 0.

### Changes made

During binding resolution, and when action state is reset, the binding index for each action is set to point to the first binding in that actions' list of bindings. Additionally, when calling InputAction.ReadValue<T>(), if the action is not currently in progress, the default value of T is run through the processors of the active binding.

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

### CODE REVIEW

File: Assets/Tests/InputSystem/CoreTests_Actions.cs [hunk 1]
```diff
@@ -29,7 +29,41 @@
...
+        action1.AddBinding("<Gamepad>/rightStick/x", processors: "normalize(min=0,max=1)");
+
+        map.Enable();
+
+        Assert.That(action1.ReadValue<float>(), Is.EqualTo(0.5f));
+    }
+
+    [Test]
+    [Category("Actions")]
+    public void Actions_WhenResetActionState_BindingIndexPointsToFirstBindingInAction()
```
Comment (Rene-Damm) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs]: The binding index aspect seems like implementation detail where the connection to the test contents isn't quite clear.

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputAction.cs [hunk 2]
```diff
@@ -980,21 +980,13 @@ object ICloneable.Clone()
...
-                    var controlIndex = actionStatePtr->controlIndex;
-                    if (controlIndex != InputActionState.kInvalidIndex)
-                        result = state.ReadValue<TValue>(actionStatePtr->bindingIndex, controlIndex);
-                }
-            }
+            if (state == null) return default(TValue);
 
-            return result;
+            var actionStatePtr = &state.actionStates[m_ActionIndexInState];
+            return phase.IsInProgress()
```
Comment (Rene-Damm) [inline Packages/com.unity.inputsystem/InputSystem/Actions/InputAction.cs]: `actionStatePtr->phase`.

Sidenote: These days, with ref vars now being supported, a whole bunch of the direct pointer code I wrote is probably better written as ref vars.

**Review by Rene-Damm — APPROVED** [review]
Some minor stuff.

File: Assets/Tests/InputSystem/CoreTests_Actions.cs (hunk 1 above)
Comment (andrew-oc) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs]: Yeah, I felt the same. I couldn't see a clearer way to test it though. It's a roundabout way of testing the binding index by using the processors attached to different bindings, which is horribly brittle in itself. Can you think of a better way?

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputAction.cs (hunk 2 above)
Comment (andrew-oc) [reply Packages/com.unity.inputsystem/InputSystem/Actions/InputAction.cs]: 👍

File: Assets/Tests/InputSystem/CoreTests_Actions.cs (hunk 1 above)
Comment (Rene-Damm) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs]: I think the test contents are fine. Just that the names seem misleading. Maybe `Actions_ReadingValueRightAfterEnabling_AppliesProcessorsFromFirstBinding` or something.

**Pull request merged**